src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -1088,6 +1088,21 @@ private HttpResponseHeaders CreateResponseHeaders(bool appCompleted)
             var hasConnection = responseHeaders.HasConnection;
             var hasTransferEncoding = responseHeaders.HasTransferEncoding;
 
+            // https://tools.ietf.org/html/rfc7540#section-8.1.2.2
+            // Such intermediaries SHOULD also remove other connection-specific header fields, such as Keep-Alive,
+            // Proxy-Connection, Transfer-Encoding, and Upgrade, even if they are not nominated by the Connection header field.
+            if (_httpVersion > Http.HttpVersion.Http11)
+            {
+                responseHeaders.ClearTransferEncoding();
+                responseHeaders.ClearConnection();
+                responseHeaders.ClearKeepAlive();
+                responseHeaders.ClearUpgrade();
+                responseHeaders.ClearProxyConnection();
+
+                hasTransferEncoding = false;
+                hasConnection = false;
```

Comment:
I don't think Kestrel should *silently* clear headers that were explicitly set by the application. Kestrel is not necessarily an "intermediary" as [defined by the spec](https://tools.ietf.org/html/draft-ietf-httpbis-semantics-12#section-3.7) which says an intermediary is an "proxy, gateway, and tunnel".

If someone is implementing an intermediary on top of Kestrel, they shouldn't copy these headers to the response. It's more efficient that way. And intermediary applications need to be careful about which response headers they copy in any case, because intermediaries need to be careful about which headers they copy to HTTP/1.x responses as well.

Assuming there's no valid reason to set these headers on a >= HTTP/2 response, I think throwing an exception when the header is set would be better. Or if that's too breaking, logging a warning.

Reply:
Kestrel should prevent the protocol violation one way or another. Removing the headers would be the friendliest, but throwing an exception would also get the message across.

Reply:
If we don't throw, I think we should log a warning at a minimum.

Reply:
I agree with @halter73 here. Logging could be noisy though

Reply:
If multiple headers are sent, do we need one log per header or one log for all?

Reply:
We should try throwing first and see if that causes too many problems.

If someone is proxying responses and naively copying these response headers without filtering them, there's probably going to be issues with HTTP/1.x as well. If someone is explicitly setting these headers to tryto get an HTTP/1.1 connection to behave a specific way, what they're trying to do won't have on an HTTP/2 client. Either way, we probably want to raise this error to developers as early and as loudly as possible.

Reply:
We should also check cross server consistency here. I bet Http.Sys filters these out.

Reply:
If we go the logging route, I would try to make it a single log containing a list of all the cleared headers.

Reply:
How about a simple debug log that says "Some of the specified response headers are prohibited for use with HTTP/2 and have been removed. See https://aka.ms/kestreldocs for details.". Listing the removed headers shouldn't be required, it will be a minimal set of well known headers. This way it only needs to be logged once per response.

Reply:
I think we should verify what other servers including IIS do before going further with this, but if we do strip headers, I don't feel a debug log is nearly in-your-face enough.

Reply:
Confirmed Http.Sys (and therefor IIS) silently filters out all 5 of these headers on HTTP/2 responses. For `transfer-encoding: chunked` it goes one step further to filter out the chunk markers from the response body.

The only header that produces an error (invalid parameter) is if `Transfer-Encoding` is set to a value other than `chunked`.

Reply:
Compromise: Clear and log at the info level.